### PR TITLE
Update road sort key values

### PR DESCRIPTION
**Pull request by rmarianski**
Connects to https://github.com/mapzen/vector-datasource/issues/546

@nvkelso could you review please?

### CODE REVIEW

**Comment by rmarianski** [comment]
I added some tests for this that might be easier to see how the values got affected: https://github.com/mapzen/tilequeue/pull/74

I know it's not the greatest place, but I put it in tilequeue because we already had an easy way for the tests to automatically run.

File: TileStache/Goodies/VecTiles/transform.py [hunk 1]
```diff
@@ -291,44 +291,45 @@ def road_classifier(shape, properties, fid, zoom):
...
     elif highway == 'secondary' or aeroway == 'runway':
-        sort_val += 20
+        sort_val += 40
     elif highway == 'tertiary' or aeroway == 'taxiway':
-        sort_val += 19
+        sort_val += 39
     elif highway.endswith('_link'):
-        sort_val += 18
+        sort_val += 38
     elif highway in ('residential', 'unclassified', 'road', 'living_street'):
```
Comment (nvkelso) [inline TileStache/Goodies/VecTiles/transform.py]: Should also include the following Natural Earth values:
`or type == 'Unknown'`

File: TileStache/Goodies/VecTiles/transform.py [hunk 2]
```diff
@@ -291,44 +291,45 @@ def road_classifier(shape, properties, fid, zoom):
...
     elif highway == 'trunk':
-        sort_val += 22
+        sort_val += 42
     elif highway == 'primary':
-        sort_val += 21
+        sort_val += 41
     elif highway == 'secondary' or aeroway == 'runway':
-        sort_val += 20
+        sort_val += 40
     elif highway == 'tertiary' or aeroway == 'taxiway':
```
Comment (nvkelso) [inline TileStache/Goodies/VecTiles/transform.py]: Should also include the following Natural Earth values:
`or type == 'Track'`

File: TileStache/Goodies/VecTiles/transform.py [hunk 3]
```diff
@@ -291,44 +291,45 @@ def road_classifier(shape, properties, fid, zoom):
...
     if highway == 'motorway':
-        sort_val += 24
+        sort_val += 44
     elif is_railway:
-        sort_val += 23
+        sort_val += 43
     elif highway == 'trunk':
-        sort_val += 22
+        sort_val += 42
     elif highway == 'primary':
```
Comment (nvkelso) [inline TileStache/Goodies/VecTiles/transform.py]: Should also include the following Natural Earth values:
`or type == 'Road'`

File: TileStache/Goodies/VecTiles/transform.py [hunk 4]
```diff
@@ -291,44 +291,45 @@ def road_classifier(shape, properties, fid, zoom):
...
+    is_railway = railway in (
+        'rail', 'tram', 'light_rail', 'narrow_guage', 'monorail')
 
     if highway == 'motorway':
-        sort_val += 24
+        sort_val += 44
     elif is_railway:
-        sort_val += 23
+        sort_val += 43
     elif highway == 'trunk':
```
Comment (nvkelso) [inline TileStache/Goodies/VecTiles/transform.py]: Should also include the following Natural Earth values:
`or type == 'Secondary Highway'`

File: TileStache/Goodies/VecTiles/transform.py [hunk 5]
```diff
@@ -291,44 +291,45 @@ def road_classifier(shape, properties, fid, zoom):
...
     railway = properties.get('railway', '')
     aeroway = properties.get('aeroway', '')
     aerialway = properties.get('aerialway', '')
     service = properties.get('service')
 
-    is_railway = railway in ('rail', 'tram', 'light_rail', 'narrow_guage', 'monorail')
+    is_railway = railway in (
+        'rail', 'tram', 'light_rail', 'narrow_guage', 'monorail')
 
     if highway == 'motorway':
```
Comment (nvkelso) [inline TileStache/Goodies/VecTiles/transform.py]: Should also include the following Natural Earth values:
`or type in ('Major Highway', 'Beltway', 'Bypass')`

File: TileStache/Goodies/VecTiles/transform.py [hunk 6]
```diff
@@ -291,44 +291,45 @@ def road_classifier(shape, properties, fid, zoom):
 def road_sort_key(shape, properties, fid, zoom):
     # Note! parse_layer_as_float must be run before this filter.
 
-    # Calculated sort value is in the range 0 to 39
-    sort_val = 0
+    floor = 300
+    ceiling = 385
```
Comment (nvkelso) [inline TileStache/Goodies/VecTiles/transform.py]: Based on ±50 for bridges and tunnels, I suspect 447 would be a better ceiling.

300 base + 5 + [50-92 for basic roads + gondolas plus spacers] + 50 + 5 = 
- _some natural buffer for things like buildings (381-50=331, and we don't allow aerialway to be tunnels)_
- road layer default (paths, steps, tracks, etc)    355
- other road service (alley, driveway)  356
- other railway service 357
- unclassified, service, minor  358
- railway yard  359
- residential, unclassified, road, living_street, Unknown   360
- railway spur, siding  361
- _reserved for water lines & other goodies 362-374_
  *_link  375
- tertiary, Track', taxiway 376
- secondary, runway 377
- primary, Road'    378
- trunk, Secondary Highway' 379
- rail, tram, light_rail, narrow_guage, monorail    380
- motorway, 'Major Highway', 'Beltway, 'Bypass' 381
- _reserved for platforms, water lines, etc 372-389_
- other aerialway   390
- chair_lift    391
- gondola, cable_car    392

File: TileStache/Goodies/VecTiles/transform.py [hunk 7]
```diff
@@ -291,44 +291,45 @@ def road_classifier(shape, properties, fid, zoom):
...
     elif aerialway == 'chair_lift':
-        sort_val += 26
-    elif aerialway != '':
-        sort_val += 25
+        sort_val += 46
+    elif aerialway:
+        sort_val += 45
     else:
-        sort_val += 15
+        sort_val += 25
```
Comment (nvkelso) [inline TileStache/Goodies/VecTiles/transform.py]: I think in this schema, the default would be `35` (not `25`) – else some tunnels will will be **above** default surface "roads"?

If we want to "pad" the ranges to allow arbitrary (client side) orders, then we should offset all up (inclusive of the default), like you're already doing. If we want more padding, we'd +10 them all again.

**Context:** We want to pad values so buildings, water strokes, transit platforms, and other features can lay above tunnels, but below roads. To do that, I think it's better to -40 for tunnels (so it's symmetrical with +40 for bridges).

We also want to give the railway values their own unique sort_val (not collide with road values).
- other road service (alley, driveway) ≠ other railway service
- unclassified, service, minor ≠ railway yard
- residential, unclassified, road, living_street, Unknown ≠ railway spur, siding

See first comment above for recommendation.

File: TileStache/Goodies/VecTiles/transform.py [hunk 8]
```diff
@@ -345,20 +346,16 @@ def road_sort_key(shape, properties, fid, zoom):
...
               (railway == 'subway' and tunnel not in ('no', 'false'))):
             sort_val -= 10
 
-        # Keep aerialways above (almost) everything else, including bridges,
-        # but make sure it doesn't go beyond the 0-34 range. (still need to
-        # leave space for explicit layer).
-        if aerialway != '':
-            sort_val = min(34, sort_val + 10)
+        if aerialway:
+            sort_val += 30
```
Comment (nvkelso) [inline TileStache/Goodies/VecTiles/transform.py]: Looking at TagInfo, only a couple dozen aerialway features have tunnel or bridge, but according to  the OSM wiki, 25% of roads with `bridge=yes` aren't tagged with a layer. To account for this we should +40 aerialway features (the same amount as for bridges) to keep aerialway always above roads. _This could be consolidated with the bridge section above?_

(The later section for `layer` order will force it to the correct data values when an OSM editor has taken the time to distinguish which way goes in what order.)

**Comment by nvkelso** [comment]
Attached is my updated OpenOffice doc (look in column c, `sort_key_v2`):

[sort_key_v2.ods.zip](https://github.com/mapzen/TileStache/files/150505/sort_key_v2.ods.zip)

**Comment by rmarianski** [comment]
@nvkelso I think I managed to incorporate all your feedback. The tests in the corresponding [tilequeue pr](https://github.com/mapzen/tilequeue/pull/74) have been updated too. You can probably more easily see what the new values for certain road types are there.

The tunnel values look to be too small now though. If the spreadsheet is the source of truth, it might be easier to try and transition the code to look more like a straight lookup, rather than a series of relative updates as it stands now.

**Comment by zerebubuth** [comment]
> If the spreadsheet is the source of truth, it might be easier to try and transition the code to look more like a straight lookup, rather than a series of relative updates as it stands now.

Yes, please! The spreadsheet is almost detailed enough that code could be generated from it. As we bring more "style" into the server, perhaps it's worth considering how to make it easier to edit. All I can think of at the moment is generalising this to a post-process step which would take a series of columns and a CSV file as configuration parameters, the CSV file having the same columns in the same order plus one for the output.

**Comment by nvkelso** [comment]
Yes, the spreadsheet should probably be the source of truth (it certainly is for the landuse values). What more should I change in the spreadsheet so it's detailed enough? Probably split out the "reserved" ranges so there's as many reserved values (rows) instead?

File: TileStache/Goodies/VecTiles/transform.py [hunk 9]
```diff
@@ -291,46 +291,50 @@ def road_classifier(shape, properties, fid, zoom):
...
     elif aerialway == 'chair_lift':
-        sort_val += 26
-    elif aerialway != '':
-        sort_val += 25
+        sort_val += 46
+    elif aerialway:
+        sort_val += 45
     else:
-        sort_val += 15
+        sort_val += 35
```
Comment (nvkelso) [inline TileStache/Goodies/VecTiles/transform.py]: 55

File: TileStache/Goodies/VecTiles/transform.py [hunk 10]
```diff
@@ -291,46 +291,50 @@ def road_classifier(shape, properties, fid, zoom):
...
     elif aerialway in ('gondola', 'cable_car'):
-        sort_val += 27
+        sort_val += 47
     elif aerialway == 'chair_lift':
-        sort_val += 26
-    elif aerialway != '':
-        sort_val += 25
+        sort_val += 46
+    elif aerialway:
+        sort_val += 45
```
Comment (nvkelso) [inline TileStache/Goodies/VecTiles/transform.py]: 90

File: TileStache/Goodies/VecTiles/transform.py [hunk 11]
```diff
@@ -291,46 +291,50 @@ def road_classifier(shape, properties, fid, zoom):
...
+    elif highway in ('service', 'minor'):
+        sort_val += 36
     elif aerialway in ('gondola', 'cable_car'):
-        sort_val += 27
+        sort_val += 47
     elif aerialway == 'chair_lift':
-        sort_val += 26
-    elif aerialway != '':
-        sort_val += 25
+        sort_val += 46
```
Comment (nvkelso) [inline TileStache/Goodies/VecTiles/transform.py]: 91

File: TileStache/Goodies/VecTiles/transform.py [hunk 12]
```diff
@@ -291,46 +291,50 @@ def road_classifier(shape, properties, fid, zoom):
...
-        sort_val += 16
+        sort_val += 38
+    elif (highway in ('residential', 'unclassified', 'road', 'living_street')
+          or ne_type == 'Unknown'):
+        sort_val += 37
+    elif highway in ('service', 'minor'):
+        sort_val += 36
     elif aerialway in ('gondola', 'cable_car'):
-        sort_val += 27
+        sort_val += 47
```
Comment (nvkelso) [inline TileStache/Goodies/VecTiles/transform.py]: 92

File: TileStache/Goodies/VecTiles/transform.py [hunk 13]
```diff
@@ -291,46 +291,50 @@ def road_classifier(shape, properties, fid, zoom):
...
 
-    is_railway = railway in ('rail', 'tram', 'light_rail', 'narrow_guage', 'monorail')
+    is_railway = railway in (
+        'rail', 'tram', 'light_rail', 'narrow_guage', 'monorail')
 
-    if highway == 'motorway':
-        sort_val += 24
+    if (highway == 'motorway' or
+            ne_type in ('Major Highway', 'Beltway', 'Bypass')):
+        sort_val += 44
```
Comment (nvkelso) [inline TileStache/Goodies/VecTiles/transform.py]: 81

File: TileStache/Goodies/VecTiles/transform.py [hunk 14]
```diff
@@ -291,46 +291,50 @@ def road_classifier(shape, properties, fid, zoom):
...
     elif is_railway:
-        sort_val += 23
-    elif highway == 'trunk':
-        sort_val += 22
-    elif highway == 'primary':
-        sort_val += 21
+        sort_val += 43
```
Comment (nvkelso) [inline TileStache/Goodies/VecTiles/transform.py]: 80

File: TileStache/Goodies/VecTiles/transform.py [hunk 15]
```diff
@@ -291,46 +291,50 @@ def road_classifier(shape, properties, fid, zoom):
...
+    elif highway == 'trunk' or ne_type == 'Secondary Highway':
+        sort_val += 42
```
Comment (nvkelso) [inline TileStache/Goodies/VecTiles/transform.py]: 79

File: TileStache/Goodies/VecTiles/transform.py [hunk 16]
```diff
@@ -291,46 +291,50 @@ def road_classifier(shape, properties, fid, zoom):
...
+    elif highway == 'primary' or ne_type == 'Road':
+        sort_val += 41
```
Comment (nvkelso) [inline TileStache/Goodies/VecTiles/transform.py]: 78

File: TileStache/Goodies/VecTiles/transform.py [hunk 17]
```diff
@@ -291,46 +291,50 @@ def road_classifier(shape, properties, fid, zoom):
...
     elif highway == 'secondary' or aeroway == 'runway':
-        sort_val += 20
-    elif highway == 'tertiary' or aeroway == 'taxiway':
-        sort_val += 19
+        sort_val += 40
```
Comment (nvkelso) [inline TileStache/Goodies/VecTiles/transform.py]: 77, though runways and highways will be styled enough differently – should they get their own sort_val?

File: TileStache/Goodies/VecTiles/transform.py [hunk 18]
```diff
@@ -291,46 +291,50 @@ def road_classifier(shape, properties, fid, zoom):
...
+    elif highway == 'tertiary' or aeroway == 'taxiway' or ne_type == 'Track':
+        sort_val += 39
```
Comment (nvkelso) [inline TileStache/Goodies/VecTiles/transform.py]: 76

File: TileStache/Goodies/VecTiles/transform.py [hunk 19]
```diff
@@ -291,46 +291,50 @@ def road_classifier(shape, properties, fid, zoom):
...
     elif highway.endswith('_link'):
-        sort_val += 18
-    elif highway in ('residential', 'unclassified', 'road', 'living_street'):
-        sort_val += 17
-    elif highway in ('unclassified', 'service', 'minor'):
-        sort_val += 16
+        sort_val += 38
```
Comment (nvkelso) [inline TileStache/Goodies/VecTiles/transform.py]: 75

File: TileStache/Goodies/VecTiles/transform.py [hunk 20]
```diff
@@ -291,46 +291,50 @@ def road_classifier(shape, properties, fid, zoom):
...
+    elif (highway in ('residential', 'unclassified', 'road', 'living_street')
+          or ne_type == 'Unknown'):
+        sort_val += 37
```
Comment (nvkelso) [inline TileStache/Goodies/VecTiles/transform.py]: This is where the reserved gap comes in:
60

File: TileStache/Goodies/VecTiles/transform.py [hunk 21]
```diff
@@ -291,46 +291,50 @@ def road_classifier(shape, properties, fid, zoom):
...
+    elif highway in ('service', 'minor'):
+        sort_val += 36
```
Comment (nvkelso) [inline TileStache/Goodies/VecTiles/transform.py]: 58

File: TileStache/Goodies/VecTiles/transform.py (hunk 9 above)
Comment (nvkelso) [inline TileStache/Goodies/VecTiles/transform.py]: With railway: 'spur', 'siding':
361 which is a -19

Comment (nvkelso) [inline TileStache/Goodies/VecTiles/transform.py]: With railway: 'yard'
359, which is a -21

Comment (nvkelso) [inline TileStache/Goodies/VecTiles/transform.py]: With railway other:
357 which is a -23

File: TileStache/Goodies/VecTiles/transform.py [hunk 22]
```diff
@@ -340,25 +344,18 @@ def road_sort_key(shape, properties, fid, zoom):
...
         sort_val -= 1
 
     if zoom >= 15:
-        # Bridges and tunnels add +/- 10
         bridge = properties.get('bridge')
         tunnel = properties.get('tunnel')
-        if bridge in ('yes', 'true'):
-            sort_val += 10
+        if bridge in ('yes', 'true') or aerialway:
+            sort_val += 40
```
Comment (nvkelso) [inline TileStache/Goodies/VecTiles/transform.py]: +50

File: TileStache/Goodies/VecTiles/transform.py [hunk 23]
```diff
@@ -340,25 +344,18 @@ def road_sort_key(shape, properties, fid, zoom):
...
         elif (tunnel in ('yes', 'true') or
               (railway == 'subway' and tunnel not in ('no', 'false'))):
-            sort_val -= 10
-
-        # Keep aerialways above (almost) everything else, including bridges,
-        # but make sure it doesn't go beyond the 0-34 range. (still need to
-        # leave space for explicit layer).
-        if aerialway != '':
-            sort_val = min(34, sort_val + 10)
+            sort_val -= 40
```
Comment (nvkelso) [inline TileStache/Goodies/VecTiles/transform.py]: -50

File: TileStache/Goodies/VecTiles/transform.py [hunk 24]
```diff
@@ -340,25 +344,18 @@ def road_sort_key(shape, properties, fid, zoom):
         else:
             sort_val -= 8
 
-    if highway == 'service' and service is not None:
+    if highway == 'service' and service:
```
Comment (nvkelso) [inline TileStache/Goodies/VecTiles/transform.py:347]: -2 (since railway would be -1 after everything else is said and done)

**Comment by nvkelso** [comment]
@rmarianski the tunnel values assume a start of 355 (not 335)?

**Comment by zerebubuth** [comment]
> Yes, the spreadsheet should probably be the source of truth (it certainly is for the landuse values). What more should I change in the spreadsheet so it's detailed enough?

We could have it behave something like this. For example, if the config was:

``` yaml
  - fn: TileStache.Goodies.VecTiles.transform.spreadsheet
    params:
      source_layer: roads
      columns: [layer, is_tunnel, is_bridge, highway, railway, service]
      output_property: sort_key
      csv: csv/roads.csv
```

And `roads.csv` was:

``` csv
layer, is_tunnel, is_bridge, highway, railway, service, sort_key
-3,*,*,*,*,*,305
-2,*,*,*,*,*,306
-1,*,*,*,*,*,307
1,*,*,*,*,*,378
2,*,*,*,*,*,379
3,*,*,*,*,*,380
*,yes,*,*,-,-,308
*,yes,*,-,*,*,309
*,yes,*,*,-,+,310
...
```

The way to interpret that would be that a feature would get the `sort_key` of the first row which matched the feature's properties, where:
1. A value matches only that value,
2. A (something)-delimited list matches any of those values,
3. A `*` matches anything, including a missing property,
4. A `+` matches anything except a missing property,
5. A `-` matches only a missing property.

**However**, a simpler approach might just be to port the Tangram style language filters, so the whole thing can be written in YAML?

@rmarianski what do you think?

**Comment by nvkelso** [comment]
Interesting, but that's sounding like new work we should undertake associated with giving every layer object (across layers) a sort_key?

**Comment by zerebubuth** [comment]
> Interesting, but that's sounding like new work we should undertake associated with giving every layer object (across layers) a sort_key?

Sure. But I think it might be considerably less error-prone than trying to manually translate spreadsheets into code. And it'll mean considerably less reviewing to try and match up the spreadsheet to the code changes.

**Comment by rmarianski** [comment]
@zerebubuth I like this idea too. With this rule based approach though, would we need to worry about performance? We're talking about sequential rule matching for each feature right? This might not actually be a problem though, because we'll have about  what, a few dozen rules? I especially like this if we can apply it as a framework for sort keys across all layers.

@nvkelso I incorporated your latest updates. Something I noticed is that now a railway with an unknown service is at the same exact level as a default road. Is that ok?

**Comment by nvkelso** [comment]
> I especially like this if we can apply it as a framework for sort keys across all layers.

+1 for across all layers.

@rmarianski I thought it would be more like 2 above default road layer (357)?

```
* road layer default (paths, steps, tracks, etc)    355
* other road service (alley, driveway)  356
* other railway service 357
* unclassified, service, minor  358
* railway yard  359
* residential, unclassified, road, living_street, Unknown   360
* railway spur, siding  361
```

**Comment by rmarianski** [comment]
The case is a railway, where highway=service, and the service tag is set to something other than spur, siding, or yard.

https://github.com/mapzen/tilequeue/blob/a2881490d01e52cb02533aca8ddc4e1f3b6963a7/tests/test_transform.py#L35-L42

This might be the code that is unexpectedly dropping it down two values lower:

https://github.com/mapzen/TileStache/blob/102281336a95d7ee45dba6182df3441525265179/TileStache/Goodies/VecTiles/transform.py#L347-L349

**Comment by zerebubuth** [comment]
> With this rule based approach though, would we need to worry about performance? We're talking about sequential rule matching for each feature right? This might not actually be a problem though, because we'll have about what, a few dozen rules?

Looks like about 223 rules from @nvkelso's original spreadsheet - about 73 for the roads layer.

It might be slow, but I don't know if it'll be that slow compared to other things we're doing (e.g: exterior boundaries, intercutting). If it is, then we'll have to look at other options, e.g: building a decision tree, JITting, byte code compiling.

I think the maintenance win makes it worthwhile, even if it turns out to be slightly slower.

**Comment by zerebubuth** [comment]
:shipit:

**Pull request merged**